src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala

```diff
@@ -0,0 +1,583 @@
+      val arxModel = AutoregressionX.fitModel(tsArray, differentialXregMatrix, p, xregMaxLag, includeOriginalXreg)
+      val ar = arxModel.coefficients.slice(0, p)
+      val xregCoef = arxModel.coefficients.slice(p, arxModel.coefficients.length + xregMaxLag)
+      val c = if (includeIntercept) Array(arxModel.c) else Array(0.0)
+
+      // [3] Calculate MA(n) part using HR
+      val ma: Array[Double] = ARIMA.hannanRissanenInit(p, q, differentialTs, includeIntercept).takeRight(q)
+
+      // [4] Fit/smooth params using CGD
+      c ++ ar ++ ma ++ xregCoef
+    } else {
+      userInitParams
+    }
+
+    val params = fitWithCSSCGD(
+        p, d, q, xregMaxLag, differentialTs, includeOriginalXreg,includeIntercept, initParams)
+
+    val model = new ARIMAXModel(p, d, q, xregMaxLag, params, includeOriginalXreg, includeIntercept)
+    model
+  }
+
+  /**
+    * Fit an ARIMAX model using conditional sum of squares estimator, optimized using conjugate
+    * gradient descent. Most of the code is copied from ARIMA fitWithCSSCGD method.
+    *
+    * @param p Autoregressive order
+    * @param d Differencing order
+    * @param q Moving average order
+    * @param xregMaxLag The maximum lag order for exogenous variables
+    * @param diffedY Differenced time series, as appropriate
+    * @param includeOriginalXreg A boolean flag indicating if the non-lagged exogenous variables should be included. Default is true.
+    * @param includeIntercept A boolean to indicate if the regression should be run without an intercept
+    * @param initParams Initial parameter guess for optimization
+    * @return Parameters optimized using CSS estimator, with method conjugate gradient descent
+    */
+  private def fitWithCSSCGD(
+                     p: Int,
+                     d: Int,
+                     q: Int,
+                     xregMaxLag: Int,
+                     diffedY: Array[Double],
+                     includeOriginalXreg: Boolean = true,
+                     includeIntercept: Boolean,
+                     initParams: Array[Double]): Array[Double] = {
+
+    val optimizer = new NonLinearConjugateGradientOptimizer(
+      NonLinearConjugateGradientOptimizer.Formula.FLETCHER_REEVES,
+      new SimpleValueChecker(1e-7, 1e-7))
+
+    val objFunction = new ObjectiveFunction(new MultivariateFunction() {
+      def value(params: Array[Double]): Double = {
+        new ARIMAXModel(p, d, q, xregMaxLag, params, includeIntercept, includeOriginalXreg).logLikelihoodCSSARMA(diffedY)
+      }
+    })
+
+    val gradient = new ObjectiveFunctionGradient(new MultivariateVectorFunction() {
+      def value(params: Array[Double]): Array[Double] = {
+        new ARIMAXModel(p, d, q, xregMaxLag, params, includeIntercept, includeOriginalXreg).gradientlogLikelihoodCSSARMA(diffedY)
+      }
+    })
+
+    val initialGuess = new InitialGuess(initParams)
+    val maxIter = new MaxIter(10000)
+    val maxEval = new MaxEval(10000)
+    val goal = GoalType.MAXIMIZE
```

Comment:
may be more readable to declare some of these inline rather than calling out separate vals.